### PR TITLE
Tree Regularizer Implementation

### PR DESCRIPTION
**Pull request by ananjan-nandi-9**
Added changes related to the tree regularizer implementation. The regularizer is the Chart class inside regularizer_new.py. The logic in the build_scores functions is mostly taken from Shikhar's original code. The get_score functions contain logic for computing tree projection scores. Also added some command line args to control the regularizer (descriptions provided along with the new arguments). Some other small additions as well: such as the functionality to dump errors from callback functions. These can be ignored since they are very rudimentary currently.

### CODE REVIEW

File: collate.py
Comment (mcgrath-k) [inline collate.py]: In general, it is best to remove commented out code for others. For this project it doesn't matter, so I won't really comment on these in the future. If you think it is a particularly helpful line, that others might want to know about, leave a description of it above :). In this case, it seems like this is just debug code since line 56-57 show what you were looking for.

File: errors/analyze.py
Comment (mcgrath-k) [inline errors/analyze.py]: Generally, the `_` character is read as a "throwaway" character. For example, imagine you want to build a list of just the values from a dictionary, and do not care about the keys:

```
value_list = [value for _, value in my_dict.items()]
```
`key` in this case would be throwaway, because we do not use it, just `value`.

In a for loop, I glance over things such as:
```for _, value in enumerate(thing):``` when doing code review, because I assume the developer wants the idx, or whatever to be thrown away.

Change this to `i` or something more descriptive `val` or `example`, etc...

File: errors/analyze.py
Comment (mcgrath-k) [inline errors/analyze.py]: Hardcoding paths should be taken note of whenever possible. Either create an argparse or make note of the hardcoding by setting the variable to all uppercase.
https://peps.python.org/pep-0008/#constants

Change to
```suggestion
    DATASET_PATH = "../data_utils/ds_addmult_mod10/data-addmult-231019.csv"
    ERROR_FILE = "/sailhome/ananjan/grokking-330/structural-grokking-330/errors/dsmult_lm_e4_d4.txt"

    train_strings = load_train(DATASET_PATH)
    error_strings = load_errors(ERROR_FILE)
```
You can put `DATASET_PATH` and `ERROR_FILE` at the top of your main function, or at the top of the file.

File: data_utils/tense_inflection_helpers.py
Comment (mcgrath-k) [inline data_utils/tense_inflection_helpers.py]: It's dangerous to copy and paste code like this (repeated in `lm_dataset_helpers.py`). Generally, any duplicated code is a bad thing across the code base, we should all look out for this.

This function is not specific to either `lm_dataset_helpers.py` or `tense_inflection_helpers.py`. In this case, I would say create a `data_utils/helpers_util.py` file that contains functions such as this that are agnostic to the helpers in the codebase.

File: data_utils/lm_dataset_helpers.py
Comment (mcgrath-k) [inline data_utils/lm_dataset_helpers.py]: Remove duplication, listed below

File: data_utils/ds_addmult_mod10_helpers.py
Comment (mcgrath-k) [inline data_utils/ds_addmult_mod10_helpers.py:1]: Let's come back to this after merge conflict is dealt with.

File: errors/analyze.py
Comment (mcgrath-k) [inline errors/analyze.py]: This functionality already exists in ds_addmult_mod10_helpers.py. Is there a reason why we cannot use the functionality in `build_dataset_addmult_mod10`?

File: models/transformer_enc_dec.py
Comment (mcgrath-k) [inline models/transformer_enc_dec.py]: Nitpick, parentheses around booleans and non-complex operations are redundant in python. If you like the visual style of this feel free to continue, but I generally recommend trying to keep things minimal / fit in with the existing style of the codebase. Don't worry about this, just an observation.

File: train_transformers.py
Comment (mcgrath-k) [inline train_transformers.py:1]: Let's come back to this after merge conflict has been resolved. Sorry @ananjan-nandi-9 this might be a bit of work due to my refactor :cry:, let me know if you want a second pair of eyes on this :)

File: training_utils.py
Comment (mcgrath-k) [inline training_utils.py]: 👍 always appreciate removing hardcoding

File: errors/analyze.py
Comment (mcgrath-k) [inline errors/analyze.py:1]: This code is a bit hard to read due to vague function names such as `get_struct` and `load_errors`, and no description of what functions are doing. It seems super useful, but without consulting with you I might not understand how to get up and running. 

Add docstrings to explain it better! If the function is unitary, a more descriptive function name is okay, but less preferred.

I highly recommend chatgpt for easily adding docstrings across the file if you have access, this should remove the hassle of creating docstrings while allowing developers to get a general sense of what is going on. Edit the generated docstring if there is something very specific happening that chatgpt did not capture, however.

The main function should especially have the docstring, since it describes what the file is doing as a whole.

Comment (mcgrath-k) [inline errors/analyze.py:1]: How can I generate the .txt file used by `load_errors`? Add a rough description of what this file is to the `main()` function docstring

File: layers/transformer/relative_transformer.py
Comment (mcgrath-k) [inline layers/transformer/relative_transformer.py]: Does this get used ever?

File: training_utils.py
Comment (mcgrath-k) [inline training_utils.py:1]: Let's come back to this after merge conflict resolve

File: regularizer_new.py
Comment (mcgrath-k) [inline regularizer_new.py:1]: This file: `regularizer_new.py` and the other semi-duplicated file `regularizer.py`, should be the main pieces of this PR to focus on.

1. First thing is to merge these two files together. Ensure that there is no duplication between these two files, which means that you will have to reconcile both the `Chart` and `Regularizer` classes (even though `Regularizer` is not present in `regularizer_new.py`). 

2. Find a good place to put the scoring functions, if `Regularizer` is a good place to house these, then pull your scoring functions in `regularizer_new.py:Chart()` into your combined file, under the `Regularizer` class. If you think that scoring can be seperated from the regularization function, then make a new class / group of functions / or utility file that is dedicated towards scoring. We can discuss the best method once you start merging.

3. Merge `build_scores_single` and `build_scores_batch`. Architecturally, these functions *could* be separated, provided that `build_scores_batch` explicitly calls `build_scores_single` to construct batches. However, this does not happen in the code. There is duplication of this functionality:

https://github.com/dcx/structural-grokking-330/pull/3/files#diff-52911badcabd9b819d1c50e22b584956534826006f51a283c6db6818ec6ffa0aR61-R87

Which can lead to unexpected mistakes when these two functions drift. Let me know if you get stuck with this and we can piece it apart and find a better solution. When things get complicated, **it is always okay to make smaller, single-responsibility functions. This also improves readability and greatly improves error prevention.**

4. Remove the duplication between the duplicated function `recurse()` in both `get_score()` and `get_score_mean()`. Make these the same function `recurse`. Then, try to find the mean outside of the recursive function, if this is not possible, it is okay to return the mean inside of the recursive function as long as you make it a flag.

5. This one is going to take time, I will only ask it for this code specifically: 
Add docstrings to explain what is happening in each function of this code. `regularizer.py` has functions that especially need this: `get_hinge` and `get_hinge_2`. These need to be combined. Unless there is an algorithm literally named Hinge Loss 2 which is functionally different from Hinge Loss: in that case it would be crucial to have docstrings explaining that there is a functional difference between the two, and why they are separate functions.


---

There are a lot of moving pieces here, and understanding how to make code extremely modular is not always immediately obvious (and takes time participating in other code reviews / being the reviewer!). Let me know if you are stuck at any point, and we can meet to go through it.

**Review by mcgrath-k — CHANGES_REQUESTED** [review]
Hi @ananjan-nandi-9, thank you so much for pulling this together!

I tried my best to keep comments on things that *will* matter greatly for keeping scientifically accurate results for this project, and allow for prototyping / casual coding where it exists. 

Other comments are general code style that have helped me keep my code modular, readable, and safer that I have learned over time through many code reviews from experts. I will mention these less on future code reviews, but I encourage you to take a look!

Pull requests are social in nature, if you have any comments / agreements / disagreements / new ideas, they are always welcome.

---
### What I expect from you
1. Read through my comments, let me know if you disagree with any piece.
2. If you agree, start applying changes detailed in my comments
3. Add commits to this branch
5. **Re-request review** from me (or @dcx !) and we can then discuss if further changes are needed, or if it all looks good and you can continue. I will close the review panels 1 by 1 if you fixed it, or respond if you have any questions.

---

As this might be your first code review of this scale: Please let me know at any point if you have questions, or if I made something unclear!

File: collate.py
Comment (ananjan-nandi-9) [reply collate.py]: Resolved, thanks!

File: data_utils/tense_inflection_helpers.py
Comment (ananjan-nandi-9) [reply data_utils/tense_inflection_helpers.py]: Resolved, thanks

File: layers/transformer/relative_transformer.py
Comment (ananjan-nandi-9) [reply layers/transformer/relative_transformer.py]: Yes, it does :-) (inside one of the wrapper interfaces)

File: models/transformer_enc_dec.py
Comment (ananjan-nandi-9) [reply models/transformer_enc_dec.py]: I prefer this style haha (habit from C++). I get the point about consistency though.

File: data_utils/lm_dataset_helpers.py
Comment (ananjan-nandi-9) [reply data_utils/lm_dataset_helpers.py]: Done

File: errors/analyze.py
Comment (ananjan-nandi-9) [reply errors/analyze.py:1]: Added recommendations

Comment (ananjan-nandi-9) [reply errors/analyze.py:1]: Done

File: errors/analyze.py
Comment (ananjan-nandi-9) [reply errors/analyze.py]: Changed

File: errors/analyze.py
Comment (ananjan-nandi-9) [reply errors/analyze.py]: There are more things in that function as well: just copied this bit for quick testing

File: errors/analyze.py
Comment (ananjan-nandi-9) [reply errors/analyze.py]: Changed

File: regularizer_new.py
Comment (ananjan-nandi-9) [reply regularizer_new.py:1]: Cleaned up this bit of the code

**Comment by ananjan-nandi-9** [comment]
First round of changes done!

File: errors/analyze.py
Comment (mcgrath-k) [reply errors/analyze.py]: We agreed in meeting to treat this as prototype, resolved

File: errors/analyze.py
Comment (mcgrath-k) [reply errors/analyze.py:1]: Thanks!

File: models/transformer_enc_dec.py
Comment (mcgrath-k) [reply models/transformer_enc_dec.py]: Understandable haha :)

File: regularizer_new.py
Comment (mcgrath-k) [reply regularizer_new.py:1]: Looks good, thanks @ananjan-nandi-9.

**Review by mcgrath-k — CHANGES_REQUESTED** [review]
Good stuff @ananjan-nandi-9 ! Let's go ahead with the merge conflict resolution and then I think we will be in good shape :)

**Review by mcgrath-k — APPROVED** [review]

**Pull request merged**
